src/Niv/Git/Cmd.hs

```diff
@@ -0,0 +1,166 @@
+import System.Process (readProcessWithExitCode)
+import qualified Data.Aeson as Aeson
+import qualified Data.HashMap.Strict as HMS
+import qualified Data.Text as T
+import qualified Options.Applicative as Opts
+import qualified Options.Applicative.Help.Pretty as Opts
+
+gitCmd :: Cmd
+gitCmd = Cmd
+  { description = describeGit
+  , parseCmdShortcut = parseGitShortcut
+  , parsePackageSpec = parseGitPackageSpec
+  , updateCmd = gitUpdate
+  , name = "git"
+  }
+
+parseGitShortcut :: T.Text -> Maybe (PackageName, Aeson.Object)
+parseGitShortcut txt'@(T.dropWhileEnd (== '/') -> txt) =
+    -- basic heuristics for figuring out if something is a git repo
+    if isGitURL
+    then case T.splitOn "/" txt of
+      [] -> Nothing
+      (last -> w) -> case T.stripSuffix ".git" w of
+        Nothing -> Just (PackageName w, HMS.singleton "repo" (Aeson.String txt'))
+        Just w' -> Just (PackageName w', HMS.singleton "repo" (Aeson.String txt'))
+    else Nothing
+  where
+    isGitURL =
+        ".git" `T.isSuffixOf` txt ||
+        "git@" `T.isPrefixOf` txt ||
+        "ssh://" `T.isPrefixOf` txt
+
+parseGitPackageSpec :: Opts.Parser PackageSpec
+parseGitPackageSpec =
+    (PackageSpec . HMS.fromList) <$>
+      many (parseRepo <|> parseRef <|> parseRev)
+  where
+    parseRepo =
+      ("repo", ) . Aeson.String <$> Opts.strOption
+        ( Opts.long "repo" <>
+          Opts.metavar "URL"
+        )
+    parseRev =
+      ("rev", ) . Aeson.String <$> Opts.strOption
+        ( Opts.long "rev" <>
+          Opts.metavar "SHA"
+        )
+    parseRef =
+      ("ref", ) . Aeson.String <$> Opts.strOption
+        ( Opts.long "ref" <>
+          Opts.metavar "REF"
+        )
+
+describeGit :: Opts.InfoMod a
+describeGit = mconcat
+  [ Opts.fullDesc
+  , Opts.progDesc "Add a git dependency. Experimental."
+  , Opts.headerDoc $ Just $
+      "Examples:" Opts.<$$>
+      "" Opts.<$$>
+      "  niv add git git@github.com:stedolan/jq" Opts.<$$>
+      "  niv add git ssh://git@github.com/stedolan/jq --rev deadb33f" Opts.<$$>
+      "  niv add git https://github.com/stedolan/jq.git" Opts.<$$>
+      "  niv add git --repo /my/custom/repo --name custom --ref foobar"
+  ]
+
+gitUpdate :: Update () ()
+gitUpdate = proc () -> do
+    useOrSet "type" -< ("git" :: Box T.Text)
+    repository <- load "repo" -< ()
+    refAndRev <- (discoverRev <+> discoverRefAndRev) -< repository
+    update "ref" -< fst <$> refAndRev
+    update "rev" -< snd <$> refAndRev
+    returnA -< ()
+  where
+    discoverRefAndRev = proc repository -> do
+      run defaultRefAndHEAD -< repository
+    discoverRev = proc repository -> do
+      ref <- load "ref" -< ()
+      rev <- run (\(r1,r2) -> latestRev r1 r2)-< (,) <$> repository <*> ref
+      returnA -< (,) <$> ref <*> rev
+
+latestRev
+    :: T.Text -- ^ the repository
+    -> T.Text -- ^ the ref/branch
+    -> IO T.Text
+latestRev repo ref = do
+    let gitArgs = [ "ls-remote", repo, "refs/heads/" <> ref ]
```

Comment:
should probably limit the number of lines we read from git; `ls-remote` can print a sh*tton of stuff